nexus/src/app/update/rot_updater.rs

```diff
@@ -0,0 +1,266 @@
+use gateway_client::types::RotSlot;
+use gateway_client::types::SpComponentFirmwareSlot;
+use gateway_client::types::SpType;
+use gateway_client::SpComponent;
+use slog::Logger;
+use std::time::Duration;
+use tokio::sync::watch;
+use uuid::Uuid;
+
+type GatewayClientError = gateway_client::Error<gateway_client::types::Error>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum RotUpdateError {
+    #[error("error communicating with MGS")]
+    MgsCommunication(#[from] GatewayClientError),
+
+    #[error("failed checking update status: {0}")]
+    PollUpdateStatus(#[from] UpdateStatusError),
+}
+
+impl From<PollUpdateStatusError> for RotUpdateError {
+    fn from(err: PollUpdateStatusError) -> Self {
+        match err {
+            PollUpdateStatusError::StatusError(err) => err.into(),
+            PollUpdateStatusError::ClientError(err) => err.into(),
+        }
+    }
+}
+
+pub struct RotUpdater {
+    log: Logger,
+    progress: watch::Sender<Option<UpdateProgress>>,
+    sp_type: SpType,
+    sp_slot: u32,
+    target_rot_slot: RotSlot,
+    update_id: Uuid,
+    // TODO-clarity maybe a newtype for this? TBD how we get this from
+    // wherever it's stored, which might give us a stronger type already.
+    rot_hubris_archive: Vec<u8>,
+}
+
+impl RotUpdater {
+    pub fn new(
+        sp_type: SpType,
+        sp_slot: u32,
+        target_rot_slot: RotSlot,
+        update_id: Uuid,
+        rot_hubris_archive: Vec<u8>,
+        log: &Logger,
+    ) -> Self {
+        let log = log.new(slog::o!(
+            "component" => "RotUpdater",
+            "sp_type" => format!("{sp_type:?}"),
+            "sp_slot" => sp_slot,
+            "target_rot_slot" => format!("{target_rot_slot:?}"),
+            "update_id" => format!("{update_id}"),
+        ));
+        let progress = watch::Sender::new(None);
+        Self {
+            log,
+            progress,
+            sp_type,
+            sp_slot,
+            target_rot_slot,
+            update_id,
+            rot_hubris_archive,
+        }
+    }
+
+    pub fn progress_watcher(&self) -> watch::Receiver<Option<UpdateProgress>> {
+        self.progress.subscribe()
+    }
+
+    /// Drive this RoT update to completion (or failure).
+    ///
+    /// Only one MGS instance is required to drive an update; however, if
+    /// multiple MGS instances are available and passed to this method and an
+    /// error occurs communicating with one instance, `RotUpdater` will try the
+    /// remaining instances before failing.
+    pub async fn update(
+        self,
+        mut mgs_clients: MgsClients,
+    ) -> Result<(), RotUpdateError> {
+        // The async blocks below want `&self` references, but we take `self`
+        // for API clarity (to start a new SP update, the caller should
+        // construct a new `SpUpdater`). Create a `&self` ref that we use
+        // through the remainder of this method.
+        let me = &self;
+
+        mgs_clients
```

Comment:
I assume it's possible that we could get a `CommunicationError` but the operation did actually happen.  What happens in that case?

Reply:
Certainly - easy case is something like "MGS got our request and sent the relevant request to the SP, then crashed before the SP responded". What happens depends on what the operation was. Going through the ones issued for an RoT update:

1. Polling for update status is a read-only operation and does not affect the SP or RoT
2. If MGS dies after starting the update but before completing it, the SP/RoT will be "stuck" waiting for the rest of the update data to be delivered, until (a) the data arrives, (b) they're reset, or (c) the update is explicitly aborted
3. Marking a target slot active - the target slot would have been marked active, but we wouldn't know; subsequent attempts to mark the same target slot active could still succeed
4. Restting the RoT - the RoT would have been reset, but we wouldn't know; the RoT would start reporting the new version (and subsequent attempts to reset it would just reset it again, without an update)

This question applies to any networked service, right? Do you think there's something different enough here that warrants more documentation?

Reply:
Yeah it applies to any networked service.  For idempotent requests it doesn't matter.  I would probably ask any time it was not obvious (to me) if the request was idempotent or that the surrounding code would handle it right.  In this case I was wondering if we would somehow get stuck (e.g., if we got a CommunicationFailed but it really succeeded, and then subsequent attempts fail because an update was in progress -- even if it was the same id).  If we handle that, great.